### PR TITLE
Constructors for String and Dict

### PR DESCRIPTION
**Pull request by musm**

### CODE REVIEW

File: src/mxarray.jl [hunk 1]
```diff
@@ -686,11 +686,10 @@ end
...
-jvalue(mx::MxArray, ty::Type{Matrix}) = jmatrix(mx)
-jvalue(mx::MxArray, ty::Type{Number}) = jscalar(mx)::Number
-jvalue(mx::MxArray, ty::Type{String}) = jstring(mx)::String
-jvalue(mx::MxArray, ty::Type{Dict}) = jdict(mx)
-jvalue(mx::MxArray, ty::Type{SparseMatrixCSC}) = jsparse(mx)
-
+convert(::Type{Array}, mx::MxArray)  = jarray(mx)
+convert(::Type{Vector}, mx::MxArray) = jvector(mx)
+convert(::Type{Matrix}, mx::MxArray) = jmatrix(mx)
+convert(::Type{Number}, mx::MxArray) = jscalar(mx)
```
Comment (simonster) [inline src/mxarray.jl]: Probably best to leave the type assert here although I'm not sure it helps much.

File: src/mxarray.jl [hunk 2]
```diff
@@ -686,11 +686,10 @@ end
...
+convert(::Type{String}, mx::MxArray) = jstring(mx)
+convert(::Type{Dict}, mx::MxArray) = jdict(mx)
+convert(::Type{SparseMatrixCSC}, mx::MxArray) = jsparse(mx)
```
Comment (yuyichao) [inline src/mxarray.jl]: As I said in the previous PR, I don't think these should be allowed to happen implicitly.

Comment (musm) [inline src/mxarray.jl]: Could you clarify the point?

Comment (yuyichao) [inline src/mxarray.jl]: The type you are converting to are too different from the type you are converting from and the operation is almost certainly something that the user should request explicitly. E.g. this is not matlab and you shouldn't construct a 1x1 array, pass it to matlab and later implicitly use the matlab array as a julia scalar.

These can be defined as constructor, which are basically "explicit" convert.

Comment (musm) [inline src/mxarray.jl]: >These can be defined as constructor, which are basically "explicit" convert.

to which function or are you saying that it was fine before with `jvalue(mx:MxArray, ::Type)`

Comment (yuyichao) [inline src/mxarray.jl]: > to which function

???

>  it was fine before with `jvalue(mx:MxArray, ::Type)`

I wasn't talking about it but yes it was fine in the sense that they are explicitly too.

File: src/mxarray.jl [hunk 3]
```diff
@@ -546,7 +547,6 @@ mxarray(d) = mxstruct(d)
 #
 ###########################################################
 
-const _mx_get_string = mxfunc(:mxGetString_730)
```
Comment (musm) [inline src/mxarray.jl]: this isn't used anywhere

File: src/mxarray.jl [hunk 4]
```diff
@@ -636,16 +636,16 @@ function jsparse(mx::MxArray)
 end
 
 
-function jstring(mx::MxArray)
+function String(mx::MxArray)
     if !(classid(mx) == mxCHAR_CLASS && ((ndims(mx) == 2 && nrows(mx) == 1) || is_empty(mx)))
-        throw(ArgumentError("jstring only applies to strings (i.e. char vectors)."))
+        throw(ArgumentError("only applies to strings (i.e. char vectors)"))
```
Comment (yuyichao) [inline src/mxarray.jl]: This doesn't look like the right way to fix the error message. It's unclear what causes the error. Same below.

File: src/mxarray.jl [hunk 5]
```diff
@@ -636,16 +636,16 @@ function jsparse(mx::MxArray)
 end
 
 
-function jstring(mx::MxArray)
+function String(mx::MxArray)
```
Comment (yuyichao) [inline src/mxarray.jl]: Needs deprecation. Same below.

Comment (musm) [inline src/mxarray.jl]: Actually I haven't deprecated `jstring` and `jdict`  I added    `jstring(mx::MxArray) = String(mx)` and similar to the bottom of the file. Later I can deprecate.

File: src/mxarray.jl [hunk 6]
```diff
@@ -546,7 +547,7 @@ mxarray(d) = mxstruct(d)
 #
 ###########################################################
 
-const _mx_get_string = mxfunc(:mxGetString_730)
+# const _mx_get_string = mxfunc(:mxGetString_730)
```
Comment (musm) [inline src/mxarray.jl]: perhaps just delete? Altought commeting out seems fine too

**Comment by musm** [comment]
I have updated and limited this PRs scope to only apply to `Dict` and `String`        the other cases are more tricky and require more thought. As far as I'm concerned, this is ready for a final review and merge.

File: test/matstr.jl [hunk 7]
```diff
@@ -70,3 +70,4 @@ $d ...
 # Test strings with =
 text = "hello = world"
 @test mat"strfind($text, 'o = w')" == 5
+
```
Comment (yuyichao) [inline test/matstr.jl]: Now this has too many new lines.

File: src/mxarray.jl [hunk 8]
```diff
@@ -234,15 +235,15 @@ function size(mx::MxArray, d::Integer)
     if d <= 0
         throw(ArgumentError("The dimension must be a positive integer."))
     end
-
```
Comment (yuyichao) [inline src/mxarray.jl]: Wasn't the conclusion that the `transcode` should stay? Even if it doesn't the change should not be in this PR.

File: src/mxarray.jl [hunk 9]
```diff
@@ -3,15 +3,15 @@
...
+
     function MxArray(p::Ptr{Void}, own::Bool)
         p == C_NULL && error("NULL pointer for MxArray.")
         self = new(p, own)
         if own
             finalizer(self, release)
         end
         return self
-    end    
+    end
```
Comment (yuyichao) [inline src/mxarray.jl]: And as I said before, these format changes make it much harder to review. (White space changes can be filtered out with `w=0` so it's marginally better than changing between `if` and `&&` or renaming variables)

File: test/matstr.jl (hunk 7 above)
Comment (musm) [inline test/matstr.jl]: ? A few PRs ago you requested I leave one newline at the end of a file.

File: src/mxarray.jl (hunk 9 above)
Comment (musm) [inline src/mxarray.jl]: ? what do you mean with `w=0` . I just trimmed trailing spaces here.

File: src/mxarray.jl (hunk 8 above)
Comment (musm) [inline src/mxarray.jl]: thanks, this was a leftover, fixed.

File: test/matstr.jl (hunk 7 above)
Comment (yuyichao) [inline test/matstr.jl]: This one now has two newlines I believe.

File: src/mxarray.jl (hunk 9 above)
Comment (yuyichao) [inline src/mxarray.jl]: > ? what do you mean with `w=0`.

https://github.com/JuliaInterop/MATLAB.jl/pull/77/files?w=0

> I just trimmed trailing spaces here.

Right. And don't do that in a PR that changes other things at the same time. It makes reviewing much harder and I almost missed the `transcode` below.

File: test/matstr.jl (hunk 7 above)
Comment (musm) [inline test/matstr.jl]: Just one, see: https://github.com/musm/MatlabCall.jl/blob/c175248c9f1bb1b16590c3f2d4c2778e7fed33b5/test/matstr.jl

**Comment by musm** [comment]
thanks again @yuyichao I really appreciate your help and review. My editor was messing up endlines. I fixed it up.

**Pull request merged**